protobuf/lib/src/protobuf/message_set.dart

```diff
@@ -0,0 +1,129 @@
+
+part of '../../protobuf.dart';
+
+const _messageSetItemsTag = 1;
+const _messageSetItemTypeIdTag = 2;
+const _messageSetItemMessageTag = 3;
+
+abstract class $_MessageSet extends GeneratedMessage {
+  @override
+  void writeToCodedBufferWriter(CodedBufferWriter output) {
+    final extensions = _fieldSet._ensureExtensions();
+
+    for (final ext in extensions._values.entries) {
+      final typeId = ext.key;
+      final message = ext.value as GeneratedMessage;
+
+      output._writeTag(_messageSetItemsTag, WIRETYPE_START_GROUP);
+      output._writeTag(_messageSetItemTypeIdTag, WIRETYPE_VARINT);
+      output._writeVarint32(typeId);
+      output._writeTag(_messageSetItemMessageTag, WIRETYPE_LENGTH_DELIMITED);
+      final mark = output._startLengthDelimited();
+      message.writeToCodedBufferWriter(output);
+      output._endLengthDelimited(mark);
+      output._writeTag(_messageSetItemsTag, WIRETYPE_END_GROUP);
+    }
+
+    final unknownFields = _fieldSet._unknownFields;
+    if (unknownFields != null) {
+      unknownFields.writeToCodedBufferWriter(output);
+    }
+  }
+
+  @override
+  void mergeFromCodedBufferReader(CodedBufferReader input,
+      [ExtensionRegistry extensionRegistry = ExtensionRegistry.EMPTY]) {
+    // Parse items. The field for the items looks like:
+    //
+    //   repeated Item items = 1;
+    //
+    // Since message sets are compatible with proto1 items can't be packed.
+    outer:
+    while (true) {
+      final tag = input.readTag();
+      final wireType = getTagWireType(tag);
+      final tagNumber = getTagFieldNumber(tag);
+
+      if (tag == 0) {
+        break;
+      }
+
+      if (tagNumber != _messageSetItemsTag) {
+        throw UnsupportedError(
+            'Invalid message set (type = $wireType, tag = $tagNumber)');
+      }
+
+      // Parse an item. An item is a message with two fields:
+      //
+      //   message Item {
+      //     int32 type_id = 2;
+      //     Message message = 3;
+      //   }
+      //
+      // We can see the fields in any order, so loop until parsing both fields.
+      int? typeId;
+      List<int>? message;
+      while (true) {
+        final tag = input.readTag();
+        final tagNumber = getTagFieldNumber(tag);
+
+        if (tag == 0) {
+          break;
+        }
+
+        if (tagNumber == _messageSetItemTypeIdTag) {
+          typeId = input.readInt32();
+          if (message != null) {
+            _parseExtension(typeId, message, extensionRegistry);
+            typeId = null;
+            message = null;
+            continue outer;
+          }
+        } else if (tagNumber == _messageSetItemMessageTag) {
+          message = input.readBytes();
+          if (typeId != null) {
+            _parseExtension(typeId, message, extensionRegistry);
+            typeId = null;
+            message = null;
+            continue outer;
+          }
+        } else {
+          throw UnsupportedError('Invalid message set item (tag = $tagNumber)');
```

Comment:
I wonder if it is better to ignore the extra field here.

I wonder what other implementations do.

Probably OK to throw. We don't have anywhere to put unknown fields.

Reply:
Yeah we could ignore the field as the encoding is basically a repeated message encoding, and when decoding messages extra fields are always OK.

OTOH if there's an extra field here it probably means some kind of non-standard behavior or a bug in the program generating the message.

No strong opinions either way..